design/kafka-upgrade/README.md

```diff
@@ -0,0 +1,175 @@
+
+Support for downgrading to an earlier version is not a priority.
+
+Upgrading Zookeeper is out of scope. We will assume that the new version of Kafka can use the existing Zookeeper version.
+
+## Upgrade in Kafka
+
+The upgrade procedure for Kafka is [documented here](https://kafka.apache.org/documentation/#upgrade).
+
+## Changes to the `Kafka` CRD
+
+Add `Kafka.spec.kafka.version` to control which version of Kafka to use.
+This corresponds to using a different image for Kafka broker pods.
+The different image will embed a different Kafka distribution.
+
+The `Kafka.spec.kafka.version` will be a string matching the pattern
+`[0-9]+(\.[0-9]+)+`. It will _not_ be constrained by the CRD to the versions 
+supported by the operator since that would require coupling CRD versions to Kafka/Operator versions.
+
+From the user's point of view the upgrade procedure is:
+
+1. Edit the `Kafka` resource setting the `version` to the version being upgraded to
+   and setting `log.message.format.version` to the version being upgraded from.
+
+    ```
+    kind: Kafka
+    spec:
+      kafka:
+        version: 2.1.0
+        config:
+          log.message.format.version: 2.0.0
+          inter.broker.protocol.version: 2.0.0
+          # ...
+    ```
+
+    This will initiate a reconciliation which upgrades the pods but not the message format or interbroker version.
+
+2. Edit the `Kafka` resource upgrading or removing the `inter.broker.protocol.version`
+
+    ```
+    kind: Kafka
+    spec:
+      kafka:
+        version: 2.1.0
+        config:
+          log.message.format.version: 2.0.0
+          # ...
+    ```
+
+    This will initiate a rolling update which means the brokers are speaking the new protocol.
+    The message format for clients is still on the old version.
+
+3. The user now has to upgrade clients. This isn't something the CO can do.
+
+4. Once the user has upgraded the clients they can update (or remove) `log.message.format.version`:
+
+    ```
+    kind: Kafka
+    spec:
+      kafka:
+        version: 2.1.0
+        config:
+          # ...
+    ```
+
+    This will initiate a final reconciliation (rolling update). The upgrade is now complete.
+
+**Note:** It will be possible to revert the upgrade upto and including step 3.
+Because `log.message.format.version` functions as an assertion that all existing messages are on or below that message version, downgrading `log.message.format.version` is insufficient to guarantee that older clients don't see messages versions they don't understand.
+This is the main reason why the CO can't fully support version downgrade.
+
+## New images and changes to the build process
+
+Supporting multiple images requires changes to the build process.
+
+A different `kafka` image itself requires a different `kafka-base`.
+
+Consequently there would logically be different versions of the images for:
+
+* `kafka-connect`
+* `kafka-connect-s2i`
+* `zookeeper`
```

Comment:
if we are saying that Zookeeper will be not update, it should always the same first old `kafka-base` right?